### PR TITLE
Balance to Asset Balance Conversion

### PR DESCRIPTION
**Pull request by apopiak**
This PR adds the `BalanceConversion` trait and `BalanceToAssetBalance` type to `pallet_assets` to allow converting native balances to asset specific ones.
This is a "low resolution" conversion based on the existential deposit (of the native currency) and the `min_balance` (of the asset).

Note: The implementation actually expects a generic fungible, so the incoming balance could theoretically be an asset as well. (So the same implementation could work for chains that only have the assets pallet.)

PR also adds some derives for `Imbalance` that seemed useful but are unconnected to the conversion.

needed for: https://github.com/paritytech/cumulus/pull/488

supersedes #8776

### CODE REVIEW

File: frame/assets/src/types.rs [hunk 1]
```diff
@@ -177,3 +181,54 @@ impl From<TransferFlags> for DebitFlags {
 		}
 	}
 }
+
+/// Converts a balance value into an asset balance.
+pub trait BalanceConversion<InBalance, AssetId, OutBalance> {
+	fn to_asset_balance(balance: InBalance, asset_id: AssetId) -> Result<OutBalance, ConversionError>;
```
Comment (kianenigma) [inline frame/assets/src/types.rs:187]:
```suggestion
	fn to_asset_balance(balance: InBalance, asset_id: AssetId) -> Result<OutBalance, Self::Error>;
```

If you expect this trait to have other external `impl`s, then it should be like this. Else, i guess it is fine.

**Review by kianenigma — COMMENTED** [review]
LGTM, but I am not really familiar with its usage on cumulus yet.

File: frame/support/src/traits/tokens/fungibles/imbalance.rs [hunk 2]
```diff
@@ -37,6 +37,7 @@ pub trait HandleImbalanceDrop<AssetId, Balance> {
 ///
 /// Importantly, it has a special `Drop` impl, and cannot be created outside of this module.
 #[must_use]
+#[derive(Debug, Eq, PartialEq)]
```
Comment (xlc) [inline frame/support/src/traits/tokens/fungibles/imbalance.rs:40]:
```suggestion
#[derive(RuntimeDebug, Eq, PartialEq)]
```

**Comment by apopiak** [comment]
I guess this needs an audit?
The particular changes are pretty innocuous, but will be relevant for transaction payments?

**Comment by apopiak** [comment]
Should the trait live in `frame_support`?

**Comment by shawntabrizi** [comment]
Probably yes, somewhere in here: https://github.com/paritytech/substrate/tree/master/frame/support/src/traits/tokens

File: frame/assets/src/types.rs [hunk 3]
```diff
@@ -177,3 +181,61 @@ impl From<TransferFlags> for DebitFlags {
 		}
 	}
 }
+
+/// Converts a balance value into an asset balance.
+pub trait BalanceConversion<InBalance, AssetId, OutBalance> {
```
Comment (kianenigma) [inline frame/assets/src/types.rs:186]: should be moved elsewhere?

**Comment by emostov** [comment]
Code itself lgtm. But what is the use case and the justification for using min_balance of assets to establish a conversion rate? From what I understand, in order to faithfully use this conversion one must assume the min_balance between two assets is equivalent in relative "value" - is this the correct way to think about it?

**Review by thiolliere — APPROVED** [review]
code is good to me,

but test need to fixed, CI doesn't compile them

**Comment by thiolliere** [comment]
> But what is the use case and the justification for using min_balance of assets to establish a conversion rate? From what I understand, in order to faithfully use this conversion one must assume the min_balance between two assets is equivalent in relative "value" - is this the correct way to think about it?

Yes I think this is the correct usecase, when min_balance hold a relatively same value.
Usually the minimum balance cover the cost to have the balance and other account information on chain.

**Review by emostov — APPROVED** [review]

**Comment by apopiak** [comment]
bot merge

**Pull request merged**